pkg/controller/operators/openshift/helpers.go

```diff
@@ -0,0 +1,234 @@
+			continue
+		}
+
+		s = append(s, skew{
+			name:                csv.GetName(),
+			namespace:           csv.GetNamespace(),
+			maxOpenShiftVersion: max.String(),
+		})
+	}
+
+	return s, utilerrors.NewAggregate(errs)
+}
+
+func desiredRelease(ctx context.Context, cli client.Client) (*semver.Version, error) {
+	cv := configv1.ClusterVersion{}
+	if err := cli.Get(ctx, client.ObjectKey{Name: "version"}, &cv); err != nil { // "version" is the name of OpenShift's ClusterVersion singleton
+		return nil, err
+	}
+
+	v := cv.Status.Desired.Version
+	if v == "" {
+		// The release version hasn't been set yet
+		return nil, nil
+	}
+
+	desired, err := semver.ParseTolerant(v)
+	if err != nil {
+		return nil, err
+	}
+
+	return &desired, nil
+}
+
+const (
+	MaxOpenShiftVersionProperty = "olm.maxOpenShiftVersion"
+)
+
+func maxOpenShiftVersion(csv *operatorsv1alpha1.ClusterServiceVersion) (*semver.Version, error) {
+	// Extract the property from the CSV's annotations if possible
+	annotation, ok := csv.GetAnnotations()[projection.PropertiesAnnotationKey]
+	if !ok {
+		return nil, nil
+	}
+
+	properties, err := projection.PropertyListFromPropertiesAnnotation(annotation)
+	if err != nil {
+		return nil, err
+	}
+
+	// Take the highest semver if there's more than one max version specified
+	var (
+		max  *semver.Version
+		dups []semver.Version
+		errs []error
+	)
+	for _, property := range properties {
+		if property.Type != MaxOpenShiftVersionProperty {
+			continue
+		}
+
+		var value string
+		if err := json.Unmarshal([]byte(property.Value), &value); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if value == "" {
+			continue
+		}
+
+		version, err := semver.ParseTolerant(value)
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		if max == nil {
+			max = &version
+			continue
+		}
+		if version.LT(*max) {
+			continue
+		}
+		if version.EQ(*max) {
+			// Found a duplicate, mark it
+			dups = append(dups, *max)
+		}
+
+		max = &version
+	}
+
+	// Return an error if THE max version has a duplicate (i.e. equivalent version)
+	// Note: This may not be a problem since there should be no difference as far as blocking upgrades is concerned.
+	// This is more for clear status messages.
+	for _, dup := range dups {
+		if max.EQ(dup) && max.String() != dup.String() { // "1.0.0" vs "1.0.0" is fine, but not "1.0.0" vs "1.0.0+1"
+			errs = append(errs, fmt.Errorf("max openshift version ambiguous, equivalent versions %s and %s have been specified concurrently", max, dup))
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+
+	return max, nil
+}
+
+func olmOperatorRelatedObjects(ctx context.Context, cli client.Client, namespace string) ([]configv1.ObjectReference, error) {
+	csvList := &operatorsv1alpha1.ClusterServiceVersionList{}
+	if err := cli.List(ctx, csvList, client.InNamespace(namespace)); err != nil {
+		return nil, err
+	}
+
+	// TODO: Is there a better way to filter (server-side maybe)?
```

Comment:
There should be an `olm.copiedFrom` label on copied CSVs. I've had this thought too, and I'm not sure why there are no examples of a CSV label selector referencing it.

Reply:
I added a label selector.